crates/manifest/src/schema/v2.rs

```diff
@@ -145,6 +187,145 @@ pub struct Component {
+    /// "foo-bar") do not map to a `ComponentDependency::Version`, or a
+    /// `ComponentDependency::Package` where the `package` is `None`.
+    fn ensure_plain_names_have_package(&self) -> anyhow::Result<()> {
+        for (dependency_name, dependency) in self.inner.iter() {
+            let DependencyName::Plain(plain) = dependency_name else {
+                continue;
+            };
+            match dependency {
+                ComponentDependency::Package { package, .. } if package.is_none() => {}
+                ComponentDependency::Version(_) => {}
+                _ => continue,
+            }
+            anyhow::bail!("dependency `{plain}` must specify a package name");
+        }
+        Ok(())
+    }
+
+    /// This method ensures that dependency names in the package form (e.g.
+    /// "foo:bar" or "foo:bar@0.1.0") do not map to specific exported
+    /// interfaces, e.g. `"foo:bar = { ..., export = "my-export" }"` is invalid.
+    fn ensure_package_names_no_export(&self) -> anyhow::Result<()> {
+        for (dependency_name, dependency) in self.inner.iter() {
+            if let DependencyName::Package(name) = dependency_name {
+                if name.interface.is_none() {
+                    let export = match dependency {
+                        ComponentDependency::Package { export, .. } => export,
+                        ComponentDependency::Local { export, .. } => export,
+                        _ => continue,
+                    };
+
+                    anyhow::ensure!(
+                        export.is_none(),
+                        "using an export to satisfy the package dependency `{dependency_name}` is not currently permitted",
+                    );
+                }
+            }
+        }
+        Ok(())
+    }
+
+    // TODO: document this invariant
+    fn ensure_disjoint(&self) -> anyhow::Result<()> {
+        for (idx, this) in self.inner.keys().enumerate() {
+            for other in self.inner.keys().skip(idx + 1) {
+                let DependencyName::Package(other) = other else {
+                    continue;
+                };
+                let DependencyName::Package(this) = this else {
+                    continue;
+                };
+
+                if this.package == other.package {
+                    Self::check_disjoint(this, other)?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn check_disjoint(
+        this: &DependencyPackageName,
+        other: &DependencyPackageName,
+    ) -> anyhow::Result<()> {
+        assert_eq!(this.package, other.package);
+
+        if let (Some(this), Some(other)) = (this.version.clone(), other.version.clone()) {
+            if Self::normalize_version(this) != Self::normalize_version(other) {
+                return Ok(());
+            }
+        }
+
+        if let (Some(this), Some(other)) = (this.interface.as_ref(), other.interface.as_ref()) {
```

Comment:
I know it is legal but I'm finding the repeated re-use of `this` and `other` to refer to different kinds of things quite distracting.  Consider e.g. `this_ver`, `this_itf`.